File: solutions/happy-number/happy-number.cpp
// @Title: Happy Number
// @Author: colinjxc
// @Date: 2021-07-17T15:06:40+08:00
// @URL: https://leetcode-cn.com/problems/happy-number


class Solution {
public:
    bool isHappy(int n) {
        int next = n;
        unordered_set<int> seen;
        
        while (next != 1) {
            next = cal(next);
            if (seen.find(next) != seen.end()) {
                return false;
            }
            seen.insert(next);
        }
        return true;
    }
                                        
private:                                        
    int cal(int n) {
        int res = 0;
        while (n > 0) {
            int nn = n % 10;
            res += nn * nn;                                                                    
            n = n / 10;
        }
        return res;
    }                                   
};

File: solutions/happy-number/happy-number.go
// @Title: Happy Number
// @Author: colinjxc
// @Date: 2021-07-17T15:25:02+08:00
// @URL: https://leetcode-cn.com/problems/happy-number


func isHappy(n int) bool {
    slow, fast := n, calc(n)
    // 快慢指针， 如果成环快的一定能追的上慢的
    for slow != 1 && slow != fast {
        slow = calc(slow) 
        fast = calc(calc(fast))                                    
    } 
    return fast == 1
}

func calc(n int) int {
    res := 0
    for n > 0 {
        res += (n%10) * (n%10)
        n = n / 10                                    
    }
    return res
}
